vaadin/flow
# feat: Incremental Gradle builds for vaadinPrepareFrontendTask

## Description

Defines the input and outputs for `vaadinPrepareFrontendTask` of Vaadin Gradle plugin making it possible to skip the task if inputs and outputs are not changed, giving `UP-TO-DATE` result, according to [incremental builds feature](https://docs.gradle.org/current/userguide/incremental_build.html#incremental_build)

Inputs are:
- path to Node.js executable
- path to npm/pnpm executable
- configuration parameters, defined in [VaadinFlowPluginExtension](https://github.com/vaadin/flow/blob/main/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt)

Outputs are:
- `package.json`, `vite.config.js` and other files generated by Vaadin in the root of the project and in the frontend folder
- frontend folder, including generated folder

according to a prepareFrontend definition here https://github.com/vaadin/flow/blob/04a69001101aec44cb43f190ca91b4b4d1491dcf/flow-plugins/flow-plugin-base/src/main/java/com/vaadin/flow/plugin/base/BuildFrontendUtil.java#L148

How to test:
- in Flow:
`git checkout cache-prepare-frontend`
`mvn clean install -DskipTests`
`cd flow-plugins/flow-gradle-plugin`
`./gradlew clean build publishToMavenLocal -x functionalTest`
- in base-starter-gradle:
add the following to `build.gradle`
```
buildscript {
    repositories {
        mavenLocal()
        ...
    }
}

repositories {
    mavenLocal()
    ...
}
```
add the following to `gradle.properties`
```
vaadinVersion=24.2-SNAPSHOT
```
run `./gradlew clean vaadinPrepareFrontend --info`

*Open questions*:
- currently there is no input or output corresponding to changes in classpath (e.g. when a JAR dependency is added, e.g. add-on or reused theme), so that the `TaskCopyFrontendFiles` might be skipped, because the whole prepare-frontend is skipped as it doesn't see changes in classpath. Not sure if it's an issue, because in dev mode this is triggered by `DevModeInitializer` and in prod mode with the `build-frontend`.
- probably it's better to remove `@CacheableTask` for V24.2 (replace it with `@DisableCachingByDefault`) and let developers enable caching manually. The way how to disable caching for a task is described [here](https://docs.gradle.org/7.6.1/userguide/more_about_tasks.html#sec:disable-state-tracking), but I couldn't make it work neither with `tasks.register` nor with `tasks.findByName` so far -> needs more testing and we have to suggest developers how to disable caching in case if Gradle false thinks that its up-to-date when it's not.
- no tests so far

Fixes https://github.com/vaadin/flow/issues/15342

## Type of change


## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.
